import "galaxy.proto";

package baidu.galaxy;

option cc_generic_services = true;

enum ProcessStatus {
    kProcessRunning = 0;
    // normal exit
    kProcessTerminate = 1;
    // by kernel killed or signal killed
    kProcessKilled = 2;
    kProcessCoreDump = 3;
}

message ProcessInfoCheckpoint {
    repeated ProcessInfo process_infos = 1;
}

message ProcessInfo {
    optional string key = 1;
    optional int32 pid = 2; 
    optional ProcessStatus status = 3;
    optional int32 exit_code = 4;
}

message GetProcessStatusRequest {
    optional string key = 1;
}

message GetProcessStatusResponse {
    optional ProcessInfo process = 1;
    optional Status status = 2;
}

message CgroupPath {
    optional string subsystem = 1;
    optional string cgroup_path = 2;
}

message ExecuteRequest {
    optional string key = 1; 
    optional string commands = 2;
    repeated string envs = 3;
    optional string path = 4;
    optional string cgroup_path = 5;
    optional string user = 6;
    optional string pty_file = 7;
    optional string chroot_path = 8;
    repeated CgroupPath cgroups = 9;
}

message ExecuteResponse {
    optional string key  = 1;
    optional int32 pid = 2;
    optional Status status = 3;
}

service Initd {
    rpc GetProcessStatus(GetProcessStatusRequest) returns (GetProcessStatusResponse);
    rpc Execute(ExecuteRequest) returns (ExecuteResponse);
}
